RecoParticleFlow/PFProducer/plugins/kdtrees/KDTreeLinkerTrackHF.cc

```diff
@@ -0,0 +1,245 @@
+    const reco::PFRecHitRef& rh = fraction[rhit].recHitRef();
+    double fract = fraction[rhit].fraction();
+
+    if ((rh.isNull()) || (fract < 1E-4))
+      continue;
+
+    const reco::PFRecHit& rechit = *rh;
+
+    // We save the links rechit to HFClusters
+    rechit2ClusterLinks_[&rechit].insert(hfCluster);
+
+    // We create a liste of rechits
+    rechitsSet_.insert(&rechit);
+  }
+}
+
+void KDTreeLinkerTrackHF::buildTree() {
+  // List of pseudo-rechits that will be used to create the KDTree
+  std::vector<KDTreeNodeInfo<reco::PFRecHit const*, 2>> eltList;
+
+  // Filling of this list
+  for (RecHitSet::const_iterator it = rechitsSet_.begin(); it != rechitsSet_.end(); it++) {
+    const reco::PFRecHit::REPPoint& posrep = (*it)->positionREP();
+
+    KDTreeNodeInfo<reco::PFRecHit const*, 2> rh1(*it, posrep.eta(), posrep.phi());
+    eltList.push_back(rh1);
+
+    // Here we solve the problem of phi circular set by duplicating some rechits
+    // too close to -Pi (or to Pi) and adding (substracting) to them 2 * Pi.
+    if (rh1.dims[1] > (M_PI - phiOffset_)) {
+      float phi = rh1.dims[1] - 2 * M_PI;
+      KDTreeNodeInfo<reco::PFRecHit const*, 2> rh2(*it, float(posrep.eta()), phi);
+      eltList.push_back(rh2);
+    }
+
+    if (rh1.dims[1] < (M_PI * -1.0 + phiOffset_)) {
+      float phi = rh1.dims[1] + 2 * M_PI;
+      KDTreeNodeInfo<reco::PFRecHit const*, 2> rh3(*it, float(posrep.eta()), phi);
+      eltList.push_back(rh3);
+    }
+  }
+
+  // Here we define the upper/lower bounds of the 2D space (eta/phi).
+  float phimin = -1.0 * M_PI - phiOffset_;
+  float phimax = M_PI + phiOffset_;
+
+  // etamin-etamax, phimin-phimax
+  KDTreeBox region(-3.0f, 3.0f, phimin, phimax);
+
+  // We may now build the KDTree
+  tree_.build(eltList, region);
+}
+
+void KDTreeLinkerTrackHF::searchLinks() {
+  // Must of the code has been taken from LinkByRecHit.cc
+
+  // We iterate over the tracks.
+  for (BlockEltSet::iterator it = targetSet_.begin(); it != targetSet_.end(); it++) {
+    reco::PFRecTrackRef trackref = (*it)->trackRefPF();
+
+    const reco::PFTrajectoryPoint& atHF = trackref->extrapolatedPoint(reco::PFTrajectoryPoint::VFcalEntrance);
+
+    // The track didn't reach hf
+    if (!atHF.isValid())
+      continue;
+
+    double dHeta = 0.;
+    float dHphi = 0.;
+    if (dHphi > M_PI)
+      dHphi = dHphi - 2. * M_PI;
+    else if (dHphi < -M_PI)
+      dHphi = dHphi + 2. * M_PI;
+
+    float tracketa = atHF.positionREP().eta() + 0.1 * dHeta;
+    float trackphi = atHF.positionREP().phi() + 0.1 * dHphi;
+
+    if (trackphi > M_PI)
+      trackphi -= 2 * M_PI;
+    else if (trackphi < -M_PI)
+      trackphi += 2 * M_PI;
+
+    // Estimate the maximal envelope in phi/eta that will be used to find rechit candidates.
+    // Same envelope for cap et barrel rechits.
+    double inflation = 1.;
+    float rangeeta = (cristalPhiEtaMaxSize_ * (1.5 + 0.5) + 0.2 * fabs(dHeta)) * inflation;
```

Comment:
there are a lot of magic numbers here. in particular, it seems that (1.5 + 0.5) could be simplified to 2.0...